src/unistd.rs

```diff
@@ -2136,3 +2136,201 @@ mod setres {
         Errno::result(res).map(drop)
     }
 }
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Passwd {
+    /// username
+    pub name: CString,
+    /// user password (probably encrypted)
+    pub passwd: CString,
+    /// user ID
+    pub uid: Uid,
+    /// group ID
+    pub gid: Gid,
+    #[cfg(not(target_os = "android"))]
+    /// user information
+    pub gecos: CString,
+    /// home directory
+    pub dir: CString,
+    /// path to shell
+    pub shell: CString,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// login class
+    pub class: CString,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// last password change
+    pub change: libc::time_t,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// expiration time of account
+    pub expire: libc::time_t
+}
+
+impl From<libc::passwd> for Passwd {
+    fn from(pw: libc::passwd) -> Passwd {
+        unsafe {
+            Passwd {
+                name: CString::new(CStr::from_ptr(pw.pw_name).to_bytes()).unwrap(),
+                passwd: CString::new(CStr::from_ptr(pw.pw_passwd).to_bytes()).unwrap(),
+                #[cfg(not(target_os = "android"))]
+                gecos: CString::new(CStr::from_ptr(pw.pw_gecos).to_bytes()).unwrap(),
+                dir: CString::new(CStr::from_ptr(pw.pw_dir).to_bytes()).unwrap(),
+                shell: CString::new(CStr::from_ptr(pw.pw_shell).to_bytes()).unwrap(),
+                uid: Uid::from_raw(pw.pw_uid),
+                gid: Gid::from_raw(pw.pw_gid),
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                class: CString::new(CStr::from_ptr(pw.pw_class).to_bytes()).unwrap(),
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                change: pw.pw_change,
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                expire: pw.pw_expire
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Group {
+    /// group name
+    pub name: CString,
+    /// group ID
+    pub gid: Gid,
+    /// list of members
+    pub mem: Vec<CString>
+}
+
+impl From<libc::group> for Group {
+    fn from(gr: libc::group) -> Group {
+        unsafe {
+            Group {
+                name: CString::new(CStr::from_ptr(gr.gr_name).to_bytes()).unwrap(),
+                gid: Gid::from_raw(gr.gr_gid),
+                mem: Group::members(gr.gr_mem)
+            }
+        }
+    }
+}
+
+impl Group {
+    unsafe fn members(mem: *mut *mut c_char) -> Vec<CString> {
+        let mut ret = vec![];
+
+        for i in 0.. {
```

Comment:
This looks like it has too many moving parts.

Reply:
I couldn't think of another way to decode a `**char`